# Curate ASVs

# Load libraries
library(tidyverse)
library(seqinr)


#### Load and format data ####
# Load sequence table generated by dada2 (ASVs inferred by pooling samples)
# (Primer sequences were trimmed in Cutadapt)
load("output/dada2/ela18s_seqtab_nochim.rda")

# Load taxonomy table generated by classifying sequences against PR2
load("output/dada2/ela18s_taxonomy_pr2v5_tryrc.rda")
#load("output/dada2/ela18s_taxonomy_pr2v412.rda")

# Format and curate taxonomy table
taxonomy <- taxonomy %>%
  as_tibble(rownames = "sequence") %>%
  mutate(across(c(domain, supergroup, division, subdivision, class, order, family, genus, species),
                ~str_remove_all(.x, ":.*"))) %>%
  mutate(across(c(supergroup, division, subdivision), ~str_remove_all(.x, "_.*")))

# Write non-redundant taxonomy to file
# taxonomy %>%
#   distinct(domain, supergroup, division, subdivision, class, order, family, genus, species) %>%
#   arrange(domain, supergroup, division, subdivision, class, order, family, genus, species) %>%
#   write_csv("output/diversity/ela18s_taxonomy_nr.csv", col_names = TRUE, na = "")


#### Assign ASV codes ####
# Number of unique ASVs
(nasvs <- n_distinct(taxonomy$sequence))

# Assign ASV codes to unique sequences
taxonomy <- taxonomy %>%
  mutate(asv_code = paste0("ASV", str_pad(string = 1:nasvs, width = nchar(nasvs), side = "left", pad = "0")))


#### Write sequences to fasta file ####
# write.fasta(sequences = as.list(taxonomy$sequence),
#             names = taxonomy$asv_code,
#             file.out = "output/fasta/ela18s_all.fasta")


#### Remove ASVs based on alignment ####
# Calculate sequence count per ASV
asv_nseqs <- seqtab.nochim %>%
  as_tibble(rownames = "sample_id") %>%
  pivot_longer(!sample_id, names_to = "sequence", values_to = "nseqs") %>%
  group_by(sequence) %>%
  summarize(nseqs = sum(nseqs))

asvs <- taxonomy %>%
  select(asv_code, sequence) %>%
  left_join(asv_nseqs, by = "sequence")

# 1. Align ASVs in MAFFT (https://mafft.cbrc.jp/alignment/server/large.html)
# 2. Visually inspect alignment and tree to identify potentially spurious ASVs

# List potentially spurious ASVs
rmseqs <-
  c("GGAAATTTTACCAGCACAGGACAGGGTCAGGATTGCCAAGTTAAAGCCTTTGCTTGATCCCCTGAAGGGTGGTGCATGGCCGCTTCTGGTTGGTGAATATGATTTGTCTGCTTAATTGCGATAACGAACGAGATCTCTGCCAATCACTAGAGCCCGCTTTCGATTTTAAGTTGGCAGTGGTTTAAGGTCGCAAGGCCTCGATCCATTGTCGCTGAGAAGGCGGAGCTTTACTTATTGGGATTATGCGGTCTACAAACGCATTGAAACAAGAGCCC",
    "GCGATCACCGAGGTTCTGCGCCATCCGGCGCGCTACGACACCTTGTACGTCGTCAACGGGGCGTTGCAGCGGCAGCCCCCGGCCGGGACGGATCCGAAGACGGTGGAGAAGGTGTACCTCGGCCACGACAAGGACGGTGCGGAGACCGGCTATGCCGTCGTGTCCGAAGCGCCGGGCTTCGCCGACCAGATCAGGGTGATCTTCGGCTATGATGCCC",
    "CGATCTACAAGGGCCTGCACATGAATACCGCCGTGCTGCTACGCAAGACGCTGCTTCAAAAACGCGACATCGTAATCATGGAAGTGCCGTATCATCTCAGCAAACACGAACCGGCAGAGGAATAGCCCTGCCGGTTCATCCAGATCGCTGTTTTATTCCTTGGGCCTGGCAAAGAATCCTGCCAGGACGCGGGCTCCATCCTGGTCTGTGATGCCC",
    "GCGATCACAGAGGTCCTGCGCCATCCGGCGCGCTACGACACGTTGTACGTCGTCAACGGGGCGTTGCAGCGGCAGCCCCCGGCCGGGACGGATCCGAAGACGGTGGAGAAGGTGTACCTCGGGCACGACAAGGACGGTGCGGAGACCGGCTATGCCGTCGTGTCCGAAGCGCCGGGCTTCGCCGACCAGATCCGGGTGATCTTCGGCTATGATGCCC",
    "GATGTCCGAGGACTTGCGCTTCGCCAGGGCGAATACCGGGCCGGCAAATGCGGCAAGCGCGGTCGCAGCGCCAGCCACAGAGGCTGCATTCATAAAACTTCTTCTGCCCATTGGTTTCATGAGGCCTCCCTATCCGATCGCGAATATGGAACCTGATACTGGGATCGGTGTGCGGAGTTTTTGCCGCCATTCTCGGCGCCGACGGGCTGTGATGCCC",
    "CAAACTTTACCAGTCCCAGATGCTATCTGGAGCAACAGATTAAGAGTTCTTTTGCGATTTGGCAGTTGGTGGTGCATGGCCGCCCCTAGTTCGTGGTTTAAACTGTCTGCTTGATTGCGATAACGAGCGAGATTGACGGGGAGAGCAACATCTCCCGAGCGCCAGCGTAGAGCTGGACGAACGCGTCAGCTCTGGCAGGTCTGTGATGCCC",
    "CAAACTTTACCAGTCCCAGATGCTATCTGGAGCAACAGATTAAGAGTTCTTTTGCGATTTGGCAGTTGGTGGTGCATGGCCGCCCCTAGTTCGTGGTTTAAACTGTCTGCTTGATTGCGATAACGAGCGAGATTGACGGGGAGAGCAACATCTCCCGAGCGCCAGCGTTGAGCTGGACGAACGCGTCAGCTCTGGCAGGTCTGTGATGCCC",
    "GAATCAATGGCGCCCGCTATGGCGAAACCGAGTGCGGCACATGCGCAGCATCCGATGTAGTTTGCGAGAAACGCCCGCGTGCGATCCATGCGATGCGACCAAATGATCATGCCGATGCAAGCCAAGAGGAAGGGGACGGTTGCGGTCAGTCCCGTTGTCGTCATTGACAAGCCGTGATGTCGAAGAACCTGCGGTAACCAGAGCGTGATGCCC",
    "GGCCACTGGAAGCCATGAATCAAATTCCAACTTGGTTGGAATTTGACGCGACGTTATCTACTGCCCCACTCTCCGGAGCTCCTACGGATATGCAGTTTGGTCTTGGCTGCGTTTAGAAAGTCCTGAAATGGTGGCCTGAAATCTCCCAAGCCATTTTCTCTAAATATTTGCAGCTTTCTTTTGATTTCCTCTGCAACGAGTCTTTTTAGTGCCAGGTCTGTGATGCCC",
    "GGAAAACTCACCAGGTCCAGAGTAATATTTGACGGACTGGTTAAAGTCCATTCTTGATGGTTACAGTAGTGGTGCATGGTCGTTTTTGGTACGTGTTGTGAAATGTCTGGTCAATTCCGTAAACGTACGAGATCTTTAACTCTAGTTCAGAAGACCTCTTGGAGGAATCTTCTCATGGCTTAGAGTAACTATATATGTACAACAATGTGGGGAAGCTGAAGGCA",
    "GATGTCCGAGGACTTGCGCTTCGCCAGGGCGAATACCGGGCCGGCAAATGCGGCAAGCGCGGTCGCAGCGCCAGCCACAGAGGCTGCATTTATAAAACTTCTTCTGCCCATTGGTTTCATGAGGCCTCCCTATCCGATCGCGAATATGGAACCTGATACTGGGATCGGTGTGCGGAGTTTTTGCCGCCATTCTTGGCGCCGACGGGCTGTGATGCCC",
    "CCCGCCCTCGTAGATGTTGATCTGATCCTGCGTGCCGTCCGGGTGCGTCGCGCTCACCTCAACTCGGCCGTCCTTTCGCGACGACGAAATCTTCGGCCGCGGCCCGTCGTTCTCCAGCGGCATCGCCAGCGTGACAATGAACTCCTCCTCGGTGCGTAAGGGCGTCCCGACGGCGGCAATCTCGATGTCGCTGATCCACAGTTCCGATTTCGGCGTGATGCCC",
    "CTCCATATGGCGGCATCCGTAGCGGCCAGGATTGGCGGGGCGTCGATCAGGATAATGTCATATTCCGCCTGGGCCTGCTTCAGGAAATCACTTACATTCTTTGAGCTGACGAGCTCAGCGGGATTAGGCGCCGTTCCCCCACCCGTGATGATATGAAGATTATCCAGGCCGGGCGTCTGCATGATCTCGTCCATGCTCATTTTCCCTGTG",
    "GTACTCCTTGTCATCAACCGTTGGATGCTCCGGGATTTCGGCTGTGCACCGTGCCGTCAGGTCCGATCTCGCCGAAATCCCGAAACACCTGCAGGTTGTTGGAGATATTCTGGGCGCCTTTGGACTTTCGACTTTTGACTTTAGACTCCGGGACTCCTTCTAGTGTCCGGCGGTGAAAACGACGCGCATGACCTCGTCCGGGGTCGTGATGCCC",
    "GCGCAGGATGTCGCGCGGGAGGCGTTTCAGGTCGCGCTCGGCGCTGGTTTCCACGGCCACCGTATAGGTCACAGGCGGCCCGCCTTTTTCAGTTCCGCTCGAATCTCATGATAGGGCCGGAACCTCTTGGCGGCGCGTCTGGCTTTGTCGAGCTCCAGCGCGTCCTCCAGGTCTTCCAGGCGCTGAATCAGGCGGCGGTAGGTCTGTGATGCCC",
    "GGAAATTTTACCAGCACAGGACAGGGTCAGGATTGCCAAGTTAAAGCCTTTGCTTGATCCCCTGAAGGGTGGTGCATGGCCGCTTCTGGTTGGTGAATGTGATTTGTCTGCTTAATTGCGATAACGAACGAGATCTCTGCCAATCACTAGAGGCCGCTTTCGCTTTTAAGCTAGTGTGTGGTTGAAGTCTTTGGCTTTTGACCTCACGCTGCTGAGAAGGCGGACCTTTACTTATTGGGATTATGCGGTCTACAAACGCATTGAAACAAGAGCCC",
    "GCGATCACAGAGGTCCTGCGCCATCCGGCGCGCTACGACACGTTGTACGTCGTCAACGGGGCGTTGCAGCGGCAGCCCCCGGCCGGGACGGATCCGAAGACGGTGGAGAAGGTGTACCTCGACCACGACAAGGACGGTGCGGAGACCGGCTATGCCGTCGTGTCCGAAGCGCCGGGCTTCGCCGACAAGATCCGGGTGATCTTCGGCTATGATGCCC",
    "CTCCATATGGCGGCATCTGTCGCGGCCAGGATTGGCGGGGCGTCGATCAGGATAATGTCATATTCCGCCTGGGCCTGCTTCAGGAAATCACTTACATTCTTTGAGCTGACGAGCTCAGCGGGATTAGGCGCCGTTCCCCCACCCGTGATGATATGAAGATTATCCAGGCCGGGCGTCTGCATGATCTCGTCCATGCTCATTTTCCCTGTG",
    "TGCTGATAATGCCTTAATTATAACTAAAGACAGCACAGGCGCAGCTGATGCAGTTACCATCACCAATGCAGGTACTGGTTGGGACATCAAGGGTACTGCAGGTGCTTGGGGTGCAACTAAGGATGGTATTATTGCTTGTACAGGTATTACAGGTACTAACCTTGTAGGTGCTGTTGGTATTCAAGGTATCACAGGATTGAATGTTCAGGGG",
    "AAAGAAGAAGAGATAAGAGGCCAGCGGAATAAAGAGCTGGTGGAAAAAATATGAATAGAAAACAATTACTTCCTAAAACCGTTAGAAATATAGAATTGCTTGCAAAAAATATTGGCATGAGTAGCCTGGATTTTTATGGACTTTTAGACCAGGGATACAAACCACACAGAGCAAAAACCAAAGGATTTTGGTTATTTGGAACTAAGAGAATTGTGATGCCC",
    "GTACTCCTTGTCATCAACCGCTGGATGTTCCGGGATTTCGGCTGTGCACTGTGCCGTCAGGTCTGATCTCACCGAAATCCCGAAACACCTGCAGGTTGTTGGGGATGTTTTGGGCGCCTTTGGACTTTCGACTTTTGACTTTAGACTCCGGGACTCCTTCTAGTGTCCGGCGGTGAAAACGACGCGCATGACCTCGTCCGGGGTCGTGATGCCC",
    "CTGGCCAAAGCCGATGAACGATTCACTACTGTCGCAGGTCTCAGAAAACATTCCCGTCAGGCTTGTACGCTGATCAAGTCGATGGCCGATGATTCCGACAAGGGCCGTGTCACCAAACGCGGTTACCTCCAGGCGGTGTTGTCCGTGATGGATAAGAAGTACAGAAGCAAGCCCAACGCCTACTACAAAATGTTTCAGACTGTGATGCCC",
    "GGCAGTTGTTTAGGCACCTTCGCCAGTTCCATGTCGGCCGTCGGATCGCGGGTTACTCTTCCCGACCGGATCAGGAACTTGAAGAGGAGCCGGACGGCGATCAGATGACTGTTCTGCGCCATGGCGCAGTAGAACTTGCCCCGTCGGGTTCGGAGTTCCGAAATGTAACGCTGGTACTCCAAGAGGACCCCGTCATCGACACACCTCAGGTCTGTGATGCCC",
    "CCGTTCTTTCATATCATCGCGTTCGACGGCGATGACAGGAAGTCCATCCACATGGAGCTCCTTGATCCTGAACTTCTCTCCCGAAGCAGGATTGACAGCGATGAACTTCTTCTTCCAAGAAGTGAAGAAGGATCCACCGTTCTCGATCTCGACACCCACAACCTTCTCATAAGGAGCTTCGTTCCCCTTAATATAGGTTGGGAACCTCCCGTCATTACCTGTGATGCCC",
    "GAAGCCAGGTGGATGGCATCGAAGCCTCTCAAAGCATGGCGGCCGATGAGCGGGGGAAGGAGTTCGTTCAGCTTTTCGTTGACCTCGACAATGATAAGACTTTTCCAGTCTTTTAGGAACGCCGCCACCAGATCCTCCATCACGGATGGATCAATGCGTTCCTCCCTGGATTTGCGGTAAAAGGCCGCTAGCAACTCGGCATAGGTGACCGCAGAGGTCTGTGATGCCC",
    "TGCGTGGCGGATCCAACGGGGAACGCCGAGCCAGGAACACTCTCCCGGATGACACAGGAGTACTCTGAAATCGCCGCGGGCAACGGCAGCATTCGTACCAGCCGGGTGCGAGACTCCGAGGGACGTGAGATGAGCGCGCATGATGCGTTCCGGCAGGTGGGGCTCGCTGAAGGCGCGCCGATTCACGCTGCGCCAACGGGAGGGTGAGTGATGCCC",
    "GTGTTCCTTTCAGTTTCCAGTTGTCAGTTTTCAGTTTGGGATGTTTCGGGATTTCGGCTTTGCACTGTGCCGTCATCTCTGATCTCACCGAAATCCCGAAACATCTGCCCCATCGACCTCGGAGCCCCCCCGACTTTCGACTCTTGACTTTCGACTCCGGGACTCCTTCTAATGTCCGGCGGTGAAGACGACCCGCATGACCTCGTCCGGGGTCGTGATGCCC",
    "ACGACATCGCCTGGTACTACATCTCTGGCTGGTATTTCGACATCGGCGCCGTCTCTAATCACAGTAGCCGTAAGAGCCGCCATCTTTTTTAGAGCGGCCATGGCTTTCTCCGAGCGGTATTCCTCTACCACACCAAGGATGACACAGGCGACCACGATAGCAAATATGACAATAGCATCGGTAATTTCCTCGGGTAGCCCCTCTCCTGGCCGATAACTGATGATGCCC",
    "GAGGCCGATGGTAAGAAGACGTCCAATCGTCTGTCTGGCCTTCGGTGAAACCTCTGTACAGCCTCGTTAATGCCCGCCATCCAGCGGTTATTAAACAGTGGGTCTGACAGGTAGCCGGCGATATCACCTATTCCGAGAGCAATCTTCGACACTCCATTGGTCGCCCAATCCTTCGTAATATCCCACCAACCCGGCTTGCCCCCGTTTTCCTCTCCAGGGAACAGGGGTTTATCAGCGGCATGCTCATTCAGGGTCCCAACGGGAGCAA",
    "GGACAACTTACCGGAGCAGGCAGCATTGTGAGAGGCGTGTGCTTAGGATGCACATGCCGAGAATGCTGCGCGTGGTGCATGGCCGTTCTTAGTTGGTGGAGCGATTTGTCTGGTTAATTCCGATAACGAACGAGACTCTAGCCTGCTAAATAGTACGCCGATCTCTTGTCGGCGCGTACTTCTTAGAGGGACAAGTAGCGGTAAGCTACACGAAATTGAGCA",
    "CGGGCGGATGACGCGGGCGACGGCCGAGAGGTTCGCAATACCTGCGGCCAGGCATTCCTCAAGCACCGGAGAATCGTGCACAACCTTCTCGACAGCAGCAGCAATGGTGATCATGTTCAAAAAATAACAATTCGTCCCAAAACAGTCAAATGATCAGAAAAATTTGACACTCCCCGTTTTTTCCCATACCCTTGGCATCACGATTTTTAAAGGGAGACTGTGATGCCC",
    "CTCACCCAGCTCCTGGCAGCCAAGTGATTTACACATGGAGAGGTAAAGCTTCACCATGTGCACCGGTACGATTTTGCCGCTTTCACTGAAAGGCTCAAGGCAATGGGTAAGCATCCATCCGCGGGCATTCCTGCTGTAACCCACTCGCTGCGGAATGGACGCCAGATAAAACACCAGCGCTGAGCTGAACGAATGTGGAAGGAGGATGCCC",
    "ATGGTGAGTCCGAATTCTAATTTGATGTTCTTCATTTTGATGATTTCTGTGTTTTTGTCGTTGGATTAGTAGTTTTCGGCAATGGCGCTGTCCTTTTGGCCGAAGCAGTTTCCGTAAAGGAATAGGCCGTCGTTGATGTCGGCGATGAACGCGAGCTTTGTCGGGTCATTCAGGCCGCGGACTTCCAGGACGCACGGAGCGACCGTGGACAGGTCTGTGATGCCC",
    "CCTGCAACAGACCGCCCGGACACCGAAGCTTTCGAGTTACACATTGTCGCCGCAGGATATCGAAGCGTTTTGGAATCATTGGCAGGCCGATGCGGCACACTACTTTCCCGACCTGCGCCGTGACGATCCACAACGTCCGACTGTGGATGCCATCTTGATCTCGCACGCTCACCTAGACCATATCGGCGACTTGGTCTATGTTGATCTGACCATCCCGTGCGTCAGCACTCGCCTGACTGCCTTTATTAGCAAAGTCTTAGTTGATATCGGAGGCAGGTCGGGCGCGCCTTTGG",
    "CGGACGCACGGGAACCGCCAGCCTAGGGGGCAGCCTCTCTGGCAAACTCACCGGGCTGGATCTCTACTGCTTCCGGCTCACGATATGCAGGTAGTCCGCGAGCGGCGCCTGCAAGTGGCCGAGTTGGGCGTGGGTGATCTCGACATCACTGGCGAGCTTCTTCTCGGCCTCAGAGAGCTCCACATCGCCGCTCTGCGACATGACGATGATCCGGTCTGTGATGCCC",
    "CACCGTGGACCTTCTTAGTACTGGCCTCGACGGTCGTCAGCCGGTCCTTGACATCCCACGACTTGTCGTTCTCGAAAACCATTCCTCCCGCGGAACCCCAGAGCGTGTTGCGATCCATGTTCAGGAGCATGAGCACCCAGGCTTGGTAAAGCGCTTCCGCCGGATAGGAGAAGTTGGCCTCGAGGCTGGCGATAGAGGCGAGGGATTCAGCCGCTTGTAGTCCGTGTTCGTTCCGGCG",
    "ATGAGGGGAGTATCTTTCTGCCGGGGAGAAGATTTCGCCAAAAATTGCTGGTTTGATCCGGCCAAACGCCCCTGGGTGGTCACTACCTCCTGCCCTTTTTCCAGAAACGCTTCCGCCAAGGTAATGCCGGCGGGCAGGTATCCCCCATGGTTCAATCGCAAATCCAAAATTAATCCGCTTAAATTGGACCCCAGTTGCTTGAAGCGTTCAGATAATTCTTGTGCGGTGTTTTCGGTAAAAATGTTTACGCGCAGGTAGCCG",
    "CGGCAATTTCGATGCACGCTGGAACAACGCTTCCGCCGAGATCTTGATGCCCGGCGTGATGGCGCCGCCCAGGTATTCTCCGTCTTTGGTCACCGCGCAGAAGGTCGTCGCGGTGCCGAAGTCCACGATGATGAGCGGCCCGCCGAAGAGCTTGAATGCAGCGGCCGCGTTCACGATGCGGTCCGCGCCCACTTCCTTGGGGTTGTCGTAGCGGATCGTGATGCCC",
    "CAGGGAAGGGAATCAGACTGGGTCGAACTTCAGGCAGATAAGGGCGCTAAATACGGCGAGGTCATCAGGATCGATCTCGGCTCAATCGAACCGATGATAGCGCAGCCCCACAGTCCCGACAACGTAGTTCCGGTGAGAGAGCTTGCAGGAACAAAAGTGCATCAGGTCTGCATCGGAAGCTGTACTAATTCTTCATATCAGATGCTCGCAACAGTCGCTTCAATGCTCAAAGGGAAATCCGTTGCAGAAGGGGTGGACCTTCTTATTAATCCCGGATC",
    "GCCAACATCCCTTATCGCAGCTCTTGCTGCCATGGTCCCGAAAGTGATTATCTGGGCGACACGATCGGAGCCGTACTTTCTCACAACATAATCAATGACTTCCTGGCGTCTCTCATAACAAAAATCTATATCTATGTCAGGCATCGAGATCCTCTCCGGATTCAAAAATCTCTCGAACATAAGCCCGTACTTTAGAGGATCAAGGCTTGTGATGCCC",
    "CTCCATATGGCGGCATCTGTCGCGGCCAGGATTGGCGGGGCGTCGATCAGGATAATGTCATATTCCGCCTGGGCCTGCTTCAGGAAATCACTTACATTCTTTGAGCTGACGAGCTCAGCGGGATTAGGCGCCGTCGTTCCCCCTGTAATGATATGGAGATTATCCAGGCCGGGCGTCTGCATGATCTCGTCCATGCTCATTTTCCCTGTG",
    "CGGCCCATTCAGGAGTGCACGGAAAAGAAATGGCGCTGGACGCAGATGACCGACGAGATCGAGTTGAAGCTTCGCCAATGGGAGATCTCGGTGCCTCTTTGCTTGAAAGCCGGCTTCTCTCCGCTCGCCGAGGAAGAGGGAACGCGGAATCCTCTTGTCTACGAGATCGATTCGTACGCGCCGGTGGTGAAGGCGATGCCGGGACCCAACCAGTACCCGGTTCCGCCCGGAAATGAGTTCGCGGTTGG",
    "GGAAATTTTACCAGCACAGGACAGGGTCAGGATTGCCAAGTTAAAGCCTTTGCTTGATCCCCTGAAGGGTGGTGCATGGCCGCTTCTGGTTGGTGAATATGATTTGTCTGCTTAATTGCGATAACGAACGAGATCTCTGCCAATCACTAGAGACCGCTTTCGATTTTAAGTTTGTGTGTGGTTGAAGTCGAAAGGCTTTGCCTCACACGGGCCTAGAAGGCGGATCTTTACTTATTGGGATTATGCGGTCTACAAACGCATTGAAACAAGAGCCC",
    "GGAAATTTTACCAGGCCAGGACGTGGCGAGGATTGCCAGATTGAAAGCTCTTGCATGATTCCGCGGAGGGTAGTGCATGGCCGCTTCTGGTTGGTGAAGTGATTTGTCTGCTTAATTGCGATAACGAACGAGATCTCTGCCAATTACTAGACAGTGCGTTCACTTTTCAGTCGGAAGCTTGGGTGTGGATTCGTCCATCCTCAAGTTCTTCGGCTGAGAACGTGCTGCTTTACTTATTGGGATTATGCATTTTTCAACTGCATTGAAACGAGAGCCC",
    "CTCCATATGGCGGCATCCGTAGCGGCCAGGATTGGCGGGGCGTCGATCAGGATAATGTCATATTCCGCCTGGGCCTGCTTCAGGAAATCACTTACATTCTTTGAGCTGACGAGCTCAGCGGGATTAGGCGCCGTTCCCCCACCCGTGATGATATGAAGATTATCCAGGCCGGGCGTCTGCATGATCTCGTCCATGCTCATTTTCCCTGTGATCAGGTCTGTGATGCC",
    "GGGCAACTTACCGGGGCAGGCAGCGTTGCGAGAAGGGCACGATTACGATTTGTGTTCTGATGATGCTGCGCGTGGTGCATGGCCGTTCTTAACACGTGGGGTGACCTGTCTGGTTAAATCCGATAACGAACGAGACTCTGTCCTGCTAAATAGATTCAGCCTTTCTATTGGGTTGAAATTTCTTCTTAGAGGGACTGGCGATTATTAGTCGCACGAGATTGAGCA",
    "CTGTTCGGATACGAGGAGGCCAAGGCCCGTATCAAAAAGATCCTCGGATCAGAAGAATTGGAAAAGAAATAACGAAAAGAACGAGGTTTACAGAGAGGCCAAAGAGAGACCCGGAGTCTATCAGTTTCCCAGTTCAGACAACTAAGTCTGGCGAAGCCGGTTAGAAACAGTTGACTCCCTCATGACGGACCGTGAGTTGCCGCAGGCATT",
    "CAACCGCAGGAATTGGGCGCGCGCCAAGAGATCAATGCAATACTCGCCGTAGTCGCCGCGCAGGGTGATCTTCTCGAAAACCATCCTTCGTGCGGCGATGAAGCCGCTGGTATAGTCACGCACGCGCCACCCTACCAGCAGCGCCGCGAACAGGTTGATTGTCACCGAAAACGCTTGTGCAAGCAAGCTGTGGCCTACGTCTTTGCCGCCCGCGACATAGCGCGATCCCAGGGCGATATCTGCGCCGCCTTCTAATCGTT",
    "CCTCAACCCCCGATGGAGGCTTCGGGCACAAGCGGTATGAAGGCCGCGCTCAACGGCATCCCGTCCCTCAGCGTCCTCGACGGATGGTGGCTCGAAGGGTGCATCGAAGGCGTGACCGGCTGGGCCATCGGCGACGACGCGAGCAAGACCACCGAGACCTCCGGCCGCTTGCGGGACGCCGCCTCGCTCTATGAGAAGCTCGGACAGGTCGTGATGCCC",
    "GACATTCTCCCTGCGCATCCATTATTGCTATCTTATAGCTATTCCGAATTGTCAACTGGATACTGGAAGATAACCTGGCGGTTTGATACACATAGCGTGCCGGCCATTGCTGATGAGCAACAGTTCTCCCTTGAATTTTTCATCAATGAGGCTTCGACCGCGGTTATTACCCAGTCCGCTCCCAAAATAATTGTCAGGGTGAAAAACATAGTTGACTGGACGGGAATTACAGGATACGGCGGG",
    "CCACTTGATCTGTCCGTCCTCGGTAGGGATCGTCACCCTCCGATGATCGTCAACGTTGTAGATGACGTATGCCGTGGCCTGCATCTGCCTGCCGTTGATCGCAATCAGGCCGATAGTCAACGTCAGTCCCGTCCCCGGCCTACGCACCCACGGCAATGGCGGAACCCAATCGGGGTGCTGTGGCTCTCCGGCAAACTCTTTGAGGTTACTGACATAGCCGTC",
    "TCCGGCATCGAGCATCAGCCGGTCGAGCATCGCCGGGGACATCCCTTTTTTCTTAGCTGATTTTACATCTTGGGTATTTGCTTTCAATATTACTCTTTTATTAGAAACCAACGCGGCGGCCATGGCGCGTAAAATCGCGTTTTTTTTATCCGGATCGAGCCCGGCAAGAAGAGTTGCGGCATCTTTTGCCGCTTTCGCAATTTTATATAGCTTATTTTTCATGCTCATTTTTGCTCCTAGAGAACCACAAGGTTATCCCGGTGCACCGCTTC",
    "AGAGTTCGTGCTGGATCCGGGTGAGTCAGAGTTGGGCGGGGGCCGGCTGGGGCGGGATGCTGATCCCGCGCATCGGGCAGGAGGTCATCGTCGAATTTCTCGAGGGCGATCCGGACTGCCCGCTCATCACCGGCCGCGTCTACAACGGTACCCATCCGCCACCGTATCCGCTGCCAGGCGAGAAGAGCAAGAGCACGCTCAAGTCGAACAGCTCACCGGATGGCGGAGGGTGCAACGAGCTGCGTTTCGAGGACAAGAAGGGGG",
    "GGAAAACTTACCGGGTCCGGACACACTGAGGATTGACAGGCGAAGTAAATATGCTAGTCCTTTCATGATTGTGTGATAGGTGATGCATGGCCGTTCTTAGTTCGTGGAGTGATCTGTCTGCTTAATTGCGTTTCAGATAAATTTGGATAATATGTAGTGACAATGTGCTTAGTGTAACAGCTAACTCATTTAACTCTATATCCATTGAAAGCAACGAACGTGACCGCAACCCTTTGTTGTGATTTAATCTATTGATTATCACAAACTAAGGGGACCGCCAGGGTTGTCAAACCGGAGGAAGGTTGCGGCA",
    "GGGCAACTTACCGGGGCAGGCGACGTAGGGTGAAGGACGTTTCTTCGCTGATCGGACTGATTACGTCGCGCGTGGTGCATGGCCGTTCTTAACACGTGGGGTGACCTGTCTGGTTAATTCCGTTAACGAACGAGACCTCCGCCTGCTAAATAGTAACAATCTTATTTATAAGGTTGGTTTATCTTCTTAGAGGGACACTGGATTTTTAATCCAGGGAAGTTGGAGGCA",
    "AGAGCCACTTTACCCGCTACAACCGGGTTATCCCCAGTCAGCATCACCGTGTGGATGGGCGGATTGATAGTCTTCAGCTGTCGCAGCGCTTCAGCACTGGTCGGGCGGGGTGCATCGGCAACCGTCACATACCCCAGCAGGCTGTCATCTTTCCCTACCAGCATCACGGTCTGTCCTTCATCCATGGCAGTTTGTATATTGGCATGTAATACCTGGTCCAGCAGGTCATCCCGCGC",
    "ATGCGCATGGATGGTGTCAACGTAGGAGCATCCGCGCGCTTGCATTCAACGCCGATACGGCGCCCGTCCTTGAACAGGAGCAGGTCGAGTTCGGCCCCGTTATGGGTGGCCCAGTAGTACGCCTCATCGGGCTGGACGGCTTTGAGGACTTCCTCGACGGCGTAGCCTTCCCACGAGGCGCCGACCTTCGGGTGATGCTCCAGGTCGTGGCGGTTGGTGATGCCC",
    "GGACAACTTACCGGAGCAGGCAGCATTGTGAGAGGCGTGCGCTTAGGATGCACATGCCGAGAATGCTGCGCGTGGTGCATGGCCGTTCTTAACACGTGGGGTGACTTGTCTGGTTAATTCCGATAACGAACGAGACTCTGTCCTGCTAACTAGTGTCCGCTTCTTTTCTGAGCGGATTATTCTTCTTAGAGGGACTGGTAGCTCTTAGCTACACGAGATTGAGCA")

# Remove potentially spurious sequences
asvs <- asvs %>%
  filter(!sequence %in% rmseqs)

# Write new fasta file for retained sequences
# write.fasta(sequences = as.list(asvs$sequence),
#             names = asvs$asv_code,
#             file.out = "output/fasta/ela18s_rmseqs.fasta")


#### Import new alignment ####
alignment_rmseqs <- read.fasta("output/fasta/ela18s_rmseqs_aligned.fasta", "DNA", as.string = TRUE, forceDNAtolower = FALSE)
alignment_rmseqs <- tibble(asv_code = names(alignment_rmseqs),
                           sequence = unlist(getSequence(alignment_rmseqs, as.string = TRUE)))

asv_start_position <- 38L
asv_end_position <- 1522L

# Trim adapter/primer positions
asv_trim <- alignment_rmseqs %>%
  mutate(rm_fwdseq = str_sub(sequence, start = 1L, end = asv_start_position - 1L),
         rm_revseq = str_sub(sequence, start = asv_end_position + 1L, end = -1L)) %>%
  mutate(rm_fwdseq = str_remove_all(rm_fwdseq, "-"),
         rm_revseq = str_remove_all(rm_revseq, "-")) %>%
  mutate(trim = case_when(rm_fwdseq != "" | (!rm_revseq %in% c("", "A", "G", "CA", "CG", "GAACA", "GAGCA", "CGGCA", "GCGGCA")) ~ TRUE,
                          TRUE ~ FALSE)) %>%
  mutate(sequence = case_when(rm_revseq %in% c("A", "G", "CA", "CG", "GAACA", "GAGCA", "CGGCA", "GCGGCA") ~ str_sub(sequence, start = asv_start_position, end = -1L),
                             TRUE ~ str_sub(sequence, start = asv_start_position, end = asv_end_position))) %>%
  mutate(sequence = str_remove_all(sequence, "-"),
         sequence = str_remove_all(sequence, "-")) %>%
  select(asv_code, sequence, trim)

# Associate updated ASVs matching existing ASVs
asv_match <- asv_trim %>%
  filter(trim == TRUE) %>%
  left_join(asvs, by = "sequence") %>%
  filter(!is.na(asv_code.y)) %>%
  select(contains("asv_code"))

# Identify redundant ASVs (i.e. matching existing ASVs)
asv_redundant <- asv_match %>%
  pull(asv_code.x)

# Identify new ASVs (after trimming primer sequences)
asv_new <- asv_trim %>%
  filter(trim == TRUE) %>%
  left_join(asvs, by = "sequence") %>%
  filter(is.na(asv_code.y)) %>%
  pull(asv_code.x)

# Some of these new ASVs are redundant with other new ASVs
seqs_new_redundant <- asv_trim %>%
  filter(asv_code %in% asv_new) %>%
  group_by(sequence) %>%
  dplyr::count() %>%
  filter(n > 1) %>%
  pull(sequence)

asv_seqs_new_redundant <- asv_trim %>%
  filter(sequence %in% seqs_new_redundant) %>%
  select(-trim)

asv_seqs_new_redundant %>%
  group_by(sequence) %>%
  dplyr::count()

asv_new_nonredundant <- asv_seqs_new_redundant %>%
  mutate(asv_code_n = as.numeric(str_remove(asv_code, "ASV"))) %>%
  group_by(sequence) %>%
  slice_min(asv_code_n) %>%
  pull(asv_code)

asv_new_redundant <- asv_seqs_new_redundant %>%
  filter(!asv_code %in% asv_new_nonredundant) %>%
  pull(asv_code)

asv_match_new <- asv_trim %>%
  filter(asv_code %in% asv_new_redundant) %>%
  left_join(asv_trim %>%
              filter(asv_code %in% asv_new_nonredundant), by = "sequence") %>%
  select(contains("asv_code"))

asv_match <- bind_rows(asv_match, asv_match_new)

# Remove redundant ASVs in set of updated, primer-trimmed ASVs
new_seqs <- asv_trim %>%
  filter(!asv_code %in% c(asv_redundant, asv_new_redundant))
length(unique(new_seqs$sequence)) == nrow(new_seqs)  # Should evaluate to TRUE (when no redundant ASV sequences)

# Write new fasta file for retrimmed retained sequences
# write.fasta(sequences = as.list(new_seqs$sequence),
#             names = new_seqs$asv_code,
#             file.out = "output/fasta/ela18s_rmseqs_trimprimers.fasta")

# Recalculate sequence counts in sequence table after collapsing redundant ASVs
seqtab_rmseqs <- seqtab.nochim %>%
  as_tibble(rownames = "sample_id") %>%
  pivot_longer(!sample_id, names_to = "sequence", values_to = "nseqs") %>%
  filter(nseqs > 0) %>%
  filter(!sequence %in% rmseqs)

seqtab_trim <- seqtab_rmseqs %>%
  left_join(asvs %>%
              select(-nseqs), by = "sequence") %>%
  left_join(asv_match, by = c("asv_code" = "asv_code.x")) %>%
  mutate(asv_code = case_when(!is.na(asv_code.y) ~ asv_code.y,
                              TRUE ~ asv_code)) %>%
  select(-asv_code.y, -sequence) %>%
  group_by(sample_id, asv_code) %>%
  summarize(nseqs = sum(nseqs)) %>%
  left_join(asv_trim %>%
              select(-trim), by = "asv_code") %>%
  ungroup()
sum(seqtab_trim$nseqs) == sum(seqtab_rmseqs$nseqs)  # Should evaluate to TRUE

# Add taxonomy information
tax_melt <- seqtab_trim %>%
  left_join(taxonomy %>%
              select(-asv_code), by = "sequence")

# Format sample information
asv_melt <- tax_melt %>%
  mutate(sample_id = str_remove(sample_id, "ELA-")) %>%
  mutate(sample_type = case_when(grepl("\\d\\d\\d\\d\\d\\d\\d\\d", sample_id) ~ "filter",
                                 TRUE ~ "sediment")) %>%
  mutate(lake_id = case_when(sample_type == "filter" ~ str_remove(sample_id, "-.*"),
                             sample_type == "sediment" ~ str_remove(sample_id, "_.*"))) %>%
  mutate(date = case_when(sample_type == "filter" ~ as.Date(str_extract(sample_id, "\\d\\d\\d\\d\\d\\d\\d\\d"), "%Y%m%d"))) %>%
  mutate(depth_unit = case_when(sample_type == "filter" ~ str_remove_all(sample_id, ".*-"),
                                sample_type == "sediment" ~ str_remove(sample_id, paste0(lake_id, "_")))) %>%
  separate(depth_unit, c("d1", "d2", "d3"), "-", remove = FALSE) %>%
  mutate(d3 = str_remove(d3, "cm")) %>%
  mutate(upperpt = case_when(d1 == "5" & d2 == "5" & d3 == "5" ~ 5,
                             d1 == "5" & d2 == "5" & d3 == "6" ~ 5.5,
                             d1 == d2 ~ as.numeric(d1),
                             TRUE ~ as.numeric(paste0(d1, ".", d2))),
         lowerpt = case_when(d1 == "5" & d2 == "5" & d3 == "5" ~ 5.5,
                             d1 == "5" & d2 == "5" & d3 == "6" ~ 6,
                             d1 == d2 ~ as.numeric(paste0(d2, ".", d3)),
                             TRUE ~ as.numeric(d3))) %>%
  select(-d1, -d2, -d3) %>%
  mutate(upperpt = case_when(sample_type == "filter" ~ as.numeric(str_remove(depth_unit, "m")),
                             TRUE ~ upperpt),
         lowerpt = case_when(sample_type == "filter" ~ as.numeric(str_remove(depth_unit, "m")),
                             TRUE ~ lowerpt)) %>%
  mutate(depth_unit = case_when(sample_type == "sediment" ~ str_c(upperpt, "-", lowerpt, "cm"),
                                sample_type == "filter" ~ depth_unit)) %>%
  mutate(sample_id = case_when(sample_type == "sediment" ~ str_c(lake_id, "_", depth_unit),
                               sample_type == "filter" ~ str_replace_all(sample_id, "-", "_"))) %>%
  mutate(midpt = case_when(sample_type == "sediment" ~ lowerpt - (lowerpt - upperpt)/2,
                           sample_type == "filter" ~ upperpt))

# Write melted sequence data to file
# asv_melt %>%
#   write_tsv("output/melted/ela18s_all_melt.tsv", col_names = TRUE)
